File: src/dataAnalytics.rb
# frozen_string_literal: true

require 'json'
require 'date'
require 'csv'
require "set"

data = JSON.parse(File.read('./data/data.json'))['games']

games_total_number = data.count

genres = data.each_with_object([]) { |game, acc| acc.concat(game['genre']) }.uniq
number_by_genre = Hash[genres.map { |genre| [genre, data.select { |game| game['genre'].include?(genre) }.count] }]

platforms = data.each_with_object([]) { |game, acc| acc.concat(game['platform']) }.uniq
number_by_platform = Hash[platforms.map { |platform| [platform, data.select { |game| game['platform'].include?(platform) }.count] }]

high_scored_games = data
                    .select { |game| game['rating']['EN']['reviews'].count > 2 }
                    .sort_by { |game| game['rating']['EN']['ratingValue'] }
                    .reverse[0..9]
                    .each_with_object({}) { |game, hash| hash[game['name']] = game['rating']['EN']['ratingValue'] }

years = data.each_with_object([]) { |game, acc| acc << Date.parse(game['dataRelized']).year }.uniq
games_per_year = Hash[years.map { |year| [year, data.select { |game| Date.parse(game['dataRelized']).year.eql?(year) }.count] }]

high_reviewed_games = data
                      .sort_by { |game| game['rating']['EN']['reviews'].count }
                      .reverse[0..9]
                      .each_with_object({}) { |game, hash| hash[game['name']] = game['rating']['EN']['reviews'].count }

authors = data.each_with_object([]) { |game, acc| acc.concat(game['author']) }.uniq
authors_games_number = Hash[authors.map { |author| [author, data.select { |game| game['author'].include?(author) }.count] }]

platform_genres_number = {}
platforms.each do |platform|
  filtered_by_platform = data.select { |game| game['platform'].include?(platform) }
  platform_genres_number[platform] = Hash[genres.map { |genre| [genre, filtered_by_platform.select { |game| game['genre'].include?(genre) }.count] }]
end

puts "Total loaded games: #{games_total_number}"

result = {
  total_number: games_total_number,
  number_by_genre: number_by_genre,
  number_by_platform: number_by_platform,
  highest_scored: high_scored_games,
  number_by_years: games_per_year,
  high_reviewed_games: high_reviewed_games,
  number_by_authors: authors_games_number,
  number_per_platform_by_genre: platform_genres_number
}

File.open("result.json","w") do |f|
  f.write(JSON.pretty_generate(result))
end

File: src/dataAnalytics.py
#!/usr/bin/env python3
import json

class DataAnalytics():

    def __init__(self, dataPathFrom, dataPathTo):
        self.dataPathFrom = dataPathFrom
        self.dataPathTo   = dataPathTo

    def countNumber(self, resultJSON, key):
        resultDict = {}
        for game in resultJSON:
            for elem in game[key]:
                if elem in resultDict:
                    resultDict[elem] += 1
                else:
                    resultDict[elem] = 0
        return resultDict

    def sortByCount(self, elem):
        return elem[1]

    def sortByValue(self, elem):
        return elem[2]

    def countYear(self, resultJSON):
        resultDict = {}
        for game in resultJSON:
            elem = game['dataRelized']
            year = int(elem.split('-')[0])
            if year in resultDict:
                resultDict[year] += 1
            else:
                resultDict[year] = 0
        return resultDict

    def platformGenres(self, resultJSON, numberPlatform):
        resultDict = {}
        for key, _ in numberPlatform.items():
            keyDict = {}
            for game in resultJSON:
                if not key in game['platform']:
                    continue
                for elem in game['genre']:
                    if elem in keyDict:
                        keyDict[elem] += 1
                    else:
                        keyDict[elem] = 0
            resultDict[key] = keyDict
        return resultDict

    def main(self):
        with open(self.dataPathFrom, 'r', encoding = 'utf-8') as file:
            resultJSON = json.loads(file.read(), strict = False)['games']

        gamesTotalNumber = len(resultJSON)

        numberGenre    = self.countNumber(resultJSON, 'genre')
        numberPlatform = self.countNumber(resultJSON, 'platform')

        resultArray = []
        for game in resultJSON:
            resultArray.append((game['name'], game['rating']['EN']['count'],
                              game['rating']['EN']['ratingValue']))
        resultArray = list(filter(lambda x: x[1] >= 10, resultArray))

        highScoreGamesList = sorted(resultArray, key = self.sortByValue)[-15:]
        highScoreGames = {elem[0]:elem[2] for elem in highScoreGamesList}

        highReviewGamesList = sorted(resultArray, key = self.sortByCount)[-15:]
        highReviewGames = {elem[0]:elem[1] for elem in highReviewGamesList}

        numberYear = self.countYear(resultJSON)

        platformGenre = self.platformGenres(resultJSON, numberPlatform)

        dataJSON = {'total_number'                :gamesTotalNumber, 
                    'number_by_genre'             :numberGenre,
                    'number_by_platform'          :numberPlatform,
                    'highest_scored'              :highScoreGames,
                    'number_by_years'             :numberYear,
                    'high_reviewed_games'         :highReviewGames,
                    'number_per_platform_by_genre':platformGenre}

        with open(self.dataPathTo, 'w', encoding = 'utf-8') as file:
            json.dump(dataJSON, file, indent = 4, ensure_ascii = False, separators = (',', ': '))

        return

if __name__ == '__main__':
    dataPathFrom = './data/data.json'
    dataPathTo   = './data/result.json'
    DataAnalytics(dataPathFrom, dataPathTo).main()
